pkg/discovery/globalnet/globalnet.go

```diff
@@ -65,152 +44,8 @@ type Config struct {
-		return 0, fmt.Errorf("%s not a valid subnet of %v", cidr, globalCidr.net)
-	}
-
-	var clusterCidr CIDR
-
-	if clusterCidr, err = NewCIDR(cidr); err != nil {
-		return 0, err
-	}
-
-	if !globalCidr.net.Contains(uintToIP(clusterCidr.lastIP)) {
-		return 0, fmt.Errorf("%s not a valid subnet of %v", cidr, globalCidr.net)
-	}
-
-	for i := 0; i < globalCidr.allocatedCount; i++ {
-		allocated := globalCidr.allocatedClusters[i]
-		if allocated.network.Contains(requestedIP) {
-			// subset of already allocated, try next
-			return allocated.lastIP, fmt.Errorf("%s subset of already allocated globalCidr %v", cidr, allocated.network)
-		}
-
-		if requestedNetwork.Contains(allocated.network.IP) {
-			// already allocated is subset of requested, no valid lastIP
-			return clusterCidr.lastIP, fmt.Errorf("%s overlaps with already allocated globalCidr %s", cidr, allocated.network)
-		}
-	}
-
-	globalCidr.allocatedClusters = append(globalCidr.allocatedClusters, &clusterCidr)
-	globalCidr.allocatedCount++
-
-	return 0, nil
-}
-
-func allocateByClusterSize(numSize uint, globalCidr *GlobalCIDR) (string, error) {
-	bitSize := bits.LeadingZeros(0) - bits.LeadingZeros(numSize-1)
-	_, totalbits := globalCidr.net.Mask.Size()
-	clusterPrefix := totalbits - bitSize
-	mask := net.CIDRMask(clusterPrefix, totalbits)
-
-	cidr := fmt.Sprintf("%s/%d", globalCidr.net.IP, clusterPrefix)
-
-	last, err := allocateByCidr(cidr, globalCidr)
-	if err != nil && last == 0 {
-		return "", err
-	}
-
-	for err != nil {
-		nextNet := net.IPNet{
-			IP:   uintToIP(last + 1),
-			Mask: mask,
-		}
-		cidr = nextNet.String()
-
-		last, err = allocateByCidr(cidr, globalCidr)
-		if err != nil && last == 0 {
-			return "", fmt.Errorf("allocation not available")
-		}
-	}
-
-	return cidr, nil
-}
-
 func AllocateGlobalCIDR(globalnetInfo *Info) (string, error) {
-	globalCidr := &GlobalCIDR{allocatedCount: 0, cidr: globalnetInfo.CidrRange}
-
-	_, network, err := net.ParseCIDR(globalCidr.cidr)
-	if err != nil {
-		return "", fmt.Errorf("invalid GlobalCIDR %s configured", globalCidr.cidr)
-	}
-
-	globalCidr.net = network
-
-	for _, globalNetwork := range globalnetInfo.CidrInfo {
-		for _, otherCluster := range globalNetwork.GlobalCIDRs {
-			otherClusterCIDR, err := NewCIDR(otherCluster)
-			if err != nil {
-				return "", err
-			}
-
-			globalCidr.allocatedClusters = append(globalCidr.allocatedClusters, &otherClusterCIDR)
-			globalCidr.allocatedCount++
-		}
-	}
-
-	return allocateByClusterSize(globalnetInfo.ClusterSize, globalCidr)
-}
-
-func ipToUint(ip net.IP) uint32 {
-	intIP := ip
-	if len(ip) == 16 {
-		intIP = ip[12:16]
-	}
-
-	return binary.BigEndian.Uint32(intIP)
-}
-
-func uintToIP(ip uint32) net.IP {
-	netIP := make(net.IP, 4)
-	binary.BigEndian.PutUint32(netIP, ip)
-
-	return netIP
+	return cidr.Allocate(&globalnetInfo.Info) //nolint:wrapcheck // No need to wrap
 }
 
 func GetValidClusterSize(cidrRange string, clusterSize uint) (uint, error) {
```

Comment:
This can also move to cidr/cidr.go as this will also get reused in clusersetip. But I can do that as part of my changes to add clustersetip code. I'd rather have this merged first.

Reply:
I didn't move this b/c I didn't think we were going to make cluster/allocation size configurable for clustersetip.